### PR TITLE
test: Cancel Mentorship Relation

### LINKED ISSUE #307: Write missing tests for Cancel Mentorship Relation API
## Description
As a developer for backend api,
I need cover most part of the logic of backend,
so that I can be confident that the app works as it is supposed to.

Write tests that cover the test cases from [this table](https://github.com/systers/mentorship-backend/wiki/Quality-Assurance#cancel-relation) for the following API:
- `PUT /mentorship_relation/{request_id}/cancel`

Also write the following tests that are not in the table:
- Valid user tries to cancel a mentorship relation with authentication token missing (FAIL)
- Valid user tries to cancel a mentorship relation with authentication token expired (FAIL)

Make sure to add the test in the appropriate [tests/mentorship_relation/test_api_cancel_relation.py](https://github.com/systers/mentorship-backend/blob/develop/tests/mentorship_relation/test_api_cancel_relation.py) file.

_this issue is intended for the Google Code-in participants, it can be released after the program ends on January 23rd_

### Update [Required]
- [ ] Write missing tests for Cancel Mentorship Relation API

## Definition of Done
- [ ] All of the required items are completed.
- [x] Approval by 1 mentor.

## Estimation
3 hours

### PR DESCRIPTION
**Pull request by mini-mini-milo**
Tested valid user cancelling valid task with missing authentication token
Tested valid user cancelling valid task with expired authentication token

### Description
Added tests for cancelling mentorship relations:
1. valid user cancelling valid task with missing authentication token (fail)
2. valid user cancelling valid task with expired authentication token (fail)
3. valid user cancelling when not involved in the relationship (fail)

Fixes #307 

### Type of Change:

- Code
- Quality Assurance

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Ran locally using "python -m unittest discover tests"

### Checklist:

- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by Tlazypanda — CHANGES_REQUESTED** [review]
Tests missing from this [table](https://github.com/systers/mentorship-backend/wiki/Quality-Assurance#cancel-relation).Can you add those tests?

**Review by isabelcosta — CHANGES_REQUESTED** [review]
Hey @mini-mini-milo if this PR is part of the GCI competition, I believe one of the steps is that this PR should be targeting the `gci-dev` branch. You can change this by clicking the edit button next to the pull request title, and change base branch from `develop` to `gci-dev`.

File: tests/mentorship_relation/test_api_cancel_relation.py [hunk 1]
```diff
@@ -61,7 +77,61 @@ def test__mentee_cancel_mentorship_relation(self):
...
+    def test__mentor_cancel_token_missing(self):
+        self.assertEqual(MentorshipRelationState.ACCEPTED, self.mentorship_relation.state)
+        with self.client:
+            response = self.client.put('/mentorship_relation/%s/cancel' % self.mentorship_relation.id) 
+            self.assertEqual(401, response.status_code)
+            self.assertEqual(MentorshipRelationState.ACCEPTED, self.mentorship_relation.state)
+            self.assertDictEqual(messages.AUTHORISATION_TOKEN_IS_MISSING, 
+                                 json.loads(response.data))
+            
+    def test__mentee_cancel_token_missing(self):
```
Comment (isabelcosta) [inline tests/mentorship_relation/test_api_cancel_relation.py]: This test and the test above are exactly equal, so this one can be deleted.
Can you please delete this?

File: tests/mentorship_relation/test_api_cancel_relation.py [hunk 2]
```diff
@@ -61,7 +77,61 @@ def test__mentee_cancel_mentorship_relation(self):
             self.assertEqual(MentorshipRelationState.CANCELLED, self.mentorship_relation.state)
             self.assertDictEqual(messages.MENTORSHIP_RELATION_WAS_CANCELLED_SUCCESSFULLY,
                              json.loads(response.data))
-
+    
+    #missing authorisation token
+    
+    def test__mentor_cancel_token_missing(self):
```
Comment (isabelcosta) [inline tests/mentorship_relation/test_api_cancel_relation.py]:
```suggestion
    def test_mentor_cancel_token_missing(self):
```

File: tests/mentorship_relation/test_api_cancel_relation.py [hunk 3]
```diff
@@ -61,7 +77,61 @@ def test__mentee_cancel_mentorship_relation(self):
...
+        with self.client:
+            response = self.client.put('/mentorship_relation/%s/cancel' % self.mentorship_relation.id) 
+            self.assertEqual(401, response.status_code)
+            self.assertEqual(MentorshipRelationState.ACCEPTED, self.mentorship_relation.state)
+            self.assertDictEqual(messages.AUTHORISATION_TOKEN_IS_MISSING, 
+                                 json.loads(response.data))
+            
+    #expired authorisation token
+            
+    def test__mentor_cancel_token_expired(self):
```
Comment (isabelcosta) [inline tests/mentorship_relation/test_api_cancel_relation.py]:
```suggestion
    def test_mentor_cancel_token_expired(self):
```

File: tests/mentorship_relation/test_api_cancel_relation.py [hunk 4]
```diff
@@ -61,7 +77,61 @@ def test__mentee_cancel_mentorship_relation(self):
...
+        self.assertEqual(MentorshipRelationState.ACCEPTED, self.mentorship_relation.state)
+        header = get_test_request_header(self.first_user.id, timedelta(days=-1))
+        with self.client:
+            response = self.client.put('/mentorship_relation/%s/cancel' % self.mentorship_relation.id, 
+                                       headers = header)
+            self.assertEqual(401, response.status_code)
+            self.assertEqual(MentorshipRelationState.ACCEPTED, self.mentorship_relation.state)
+            self.assertDictEqual(messages.TOKEN_HAS_EXPIRED, json.loads(response.data))
+            
+    def test__mentee_cancel_token_expired(self):
```
Comment (isabelcosta) [inline tests/mentorship_relation/test_api_cancel_relation.py]:
```suggestion
    def test_mentee_cancel_token_expired(self):
```

File: tests/test_data.py [hunk 5]
```diff
@@ -23,3 +23,12 @@
...
 }
+
+
+user3 = {
+    "name": "Userc",
+    "email": "user3@email.com",
+    "username": "user3",
+    "password": "user3_pwd",
+    "terms_and_conditions_checked": True
+}
```
Comment (isabelcosta) [inline tests/test_data.py]: Please add a blank line here in the end of the file

File: tests/mentorship_relation/test_api_cancel_relation.py [hunk 6]
```diff
@@ -61,7 +77,61 @@ def test__mentee_cancel_mentorship_relation(self):
...
+        with self.client:
+            response = self.client.put('/mentorship_relation/%s/cancel' % self.mentorship_relation.id, 
+                                       headers = header)
+            self.assertEqual(401, response.status_code)
+            self.assertEqual(MentorshipRelationState.ACCEPTED, self.mentorship_relation.state)
+            self.assertDictEqual(messages.TOKEN_HAS_EXPIRED, json.loads(response.data))
+        
+    #not involved in mentorship    
+        
+    def test__user_not_in_mentorship(self):
```
Comment (isabelcosta) [inline tests/mentorship_relation/test_api_cancel_relation.py]:
```suggestion
    def test_user_not_involved_tries_to_cancel_relation(self):
```

**Review by isabelcosta — CHANGES_REQUESTED** [review]
Hey @mini-mini-milo I left some request changes. Its small things, you are almost done!
Thank you for your contribution 🙌

File: tests/mentorship_relation/test_api_cancel_relation.py (hunk 2 above)
Comment (isabelcosta) [reply tests/mentorship_relation/test_api_cancel_relation.py]: I know there are other tests in this file with double `_`, it must have been a typo that stayed. we can fix it later.

**Review by isabelcosta — CHANGES_REQUESTED** [review]
You deleted one test that was ok, I just asked you to remove one of the tests which was `test__mentor_cancel_token_missing`.
You also removed `test__mentee_cancel_token_expired` test, that was needed. Can you please put that test again?

**Review by Tlazypanda — APPROVED** [review]
Approved :)

**Comment by isabelcosta** [comment]
Thank you for your contribution here @mini-mini-milo  🎉 
I'm going to close this PR, since we decided to merge another one.

**Pull request closed without merging**